File: samples/shared_camera_java/app/src/main/java/com/google/ar/core/examples/java/common/rendering/Texture.java
/*
 * Copyright 2020 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.ar.core.examples.java.common.rendering;

import static android.opengl.GLES20.GL_CLAMP_TO_EDGE;
import static android.opengl.GLES20.GL_TEXTURE_2D;
import static android.opengl.GLES20.GL_TEXTURE_MAG_FILTER;
import static android.opengl.GLES20.GL_TEXTURE_MIN_FILTER;
import static android.opengl.GLES20.GL_TEXTURE_WRAP_S;
import static android.opengl.GLES20.GL_TEXTURE_WRAP_T;
import static android.opengl.GLES20.GL_UNSIGNED_BYTE;
import static android.opengl.GLES20.glBindTexture;
import static android.opengl.GLES20.glGenTextures;
import static android.opengl.GLES20.glTexImage2D;
import static android.opengl.GLES20.glTexParameteri;
import static android.opengl.GLES30.GL_LINEAR;
import static android.opengl.GLES30.GL_RG;
import static android.opengl.GLES30.GL_RG8;

import android.media.Image;
import com.google.ar.core.Frame;
import com.google.ar.core.exceptions.NotYetAvailableException;

/** Handle the creation and update of a GPU texture. */
public final class Texture {
  // Stores the latest provided texture id.
  private int textureId = -1;
  private int width = -1;
  private int height = -1;

  /**
   * Creates and initializes the texture. This method needs to be called on a thread with a EGL
   * context attached.
   */
  public void createOnGlThread() {
    int[] textureIdArray = new int[1];
    glGenTextures(1, textureIdArray, 0);
    textureId = textureIdArray[0];
    glBindTexture(GL_TEXTURE_2D, textureId);
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
  }

  /**
   * Updates the texture with the content from acquireDepthImage, which provides an image in DEPTH16
   * format, representing each pixel as a depth measurement in millimeters. This method needs to be
   * called on a thread with a EGL context attached.
   */
  public void updateWithDepthImageOnGlThread(final Frame frame) {
    try {
      Image depthImage = frame.acquireDepthImage();
      width = depthImage.getWidth();
      height = depthImage.getHeight();
      glBindTexture(GL_TEXTURE_2D, textureId);
      glTexImage2D(
          GL_TEXTURE_2D,
          0,
          GL_RG8,
          width,
          height,
          0,
          GL_RG,
          GL_UNSIGNED_BYTE,
          depthImage.getPlanes()[0].getBuffer());
      depthImage.close();
    } catch (NotYetAvailableException e) {
      // This normally means that depth data is not available yet. This is normal so we will not
      // spam the logcat with this.
    }
  }

  public int getTextureId() {
    return textureId;
  }

  public int getWidth() {
    return width;
  }

  public int getHeight() {
    return height;
  }
}

File: samples/hardwarebuffer_c/app/src/main/cpp/texture.cc
/*
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "texture.h"

// clang-format off
#include <GLES3/gl31.h>
#include <GLES2/gl2ext.h>
// clang-format on
#include "util.h"

namespace hello_ar {

void Texture::CreateOnGlThread() {
  GLuint texture_id_array[1];
  glGenTextures(1, texture_id_array);
  texture_id_ = texture_id_array[0];

  glBindTexture(GL_TEXTURE_2D, texture_id_);
  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
}

void Texture::UpdateWithDepthImageOnGlThread(const ArSession& session,
                                             const ArFrame& frame) {
  ArImage* depth_image = nullptr;
  if (ArFrame_acquireDepthImage16Bits(&session, &frame, &depth_image) !=
      AR_SUCCESS) {
    // No depth image received for this frame.
    return;
  }
  // Checks that the format is as expected.
  ArImageFormat image_format;
  ArImage_getFormat(&session, depth_image, &image_format);
  if (image_format != AR_IMAGE_FORMAT_D_16) {
    LOGE("Unexpected image format 0x%x", image_format);
    ArImage_release(depth_image);
    abort();
    return;
  }

  const uint8_t* depth_data = nullptr;
  int plane_size_bytes = 0;
  ArImage_getPlaneData(&session, depth_image, /*plane_index=*/0, &depth_data,
                       &plane_size_bytes);

  // Bails out if there's no depth_data.
  if (depth_data == nullptr) {
    ArImage_release(depth_image);
    return;
  }

  // Sets texture sizes.
  int image_width = 0;
  int image_height = 0;
  int image_pixel_stride = 0;
  int image_row_stride = 0;
  ArImage_getWidth(&session, depth_image, &image_width);
  ArImage_getHeight(&session, depth_image, &image_height);
  ArImage_getPlanePixelStride(&session, depth_image, 0, &image_pixel_stride);
  ArImage_getPlaneRowStride(&session, depth_image, 0, &image_row_stride);
  ArImage_release(depth_image);
  glBindTexture(GL_TEXTURE_2D, texture_id_);
  glTexImage2D(GL_TEXTURE_2D, 0, GL_RG8, image_width, image_height, 0, GL_RG,
               GL_UNSIGNED_BYTE, depth_data);
  width_ = image_width;
  height_ = image_height;
}

}  // namespace hello_ar
